src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/ConnectionLifetime.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Adapter;
+using Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Transport;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal
+{
+    public class ConnectionLifetime : IConnectionContext
```

Comment:
I think this class should be named `Connection`, or better `HttpConnection` to avoid ambiguity with transport connection classes.

The reason I think this one shouldn't have `Lifetime` in its name is that it does more than lifetime stuff e.g. exposes I/O stuff. OTOH the other class exposes just `End`, which is clearly related to the connection's lifetime.